### DISCUSSION TITLE
How to get the avarage value of variable at the quadrature points in an element?

### DISCUSSION THREAD

**Discussion by xiaojbing** [idaholab/moose#22365] [Q&A General] [answered]
I want to get the variable value on all Gaussian quadrature points (`_q_point[_qp]`) inside an element, and get the average value of them. Is there any way to distinguish the elements where different Gaussian integration points (_q_point[_qp]) are located? Or, is there any function to get the numbered values of all grid elements and iterate over all the numbers to retrieve the values on the Gaussian integration points (`_q_point[_qp]`)?
In other words, how do I determine what cell number (_current_elem -> id()) corresponds to _qp when it is a definite number?

**Comment by GiudGiud** [comment, marked as answer]
You cant match from _qp to the cell_id, _qp is the local index of the quadrature points when looping over elements. 

The loops over all elements are built in ElementalUserObjects, and some GeneralUserObjects are doing this on their own.

Are you trying to project from a higher order variable to a CONST MONOMIAL that will be constant on each element?
If so, I d make a multiapp with a projection transfer then transfer that back to the main app

**Reply by xiaojbing** [reply]
Thank you, Can Auxkernel do this projection?

**Reply by GiudGiud** [reply]
the SelfAux is the closest to that but I dont think it will do the projection properly

**Reply by GiudGiud** [reply]
check the code before you use it

**Reply by xiaojbing** [reply]
Thank you. From high order to low order, I recently tried to implement it in the input file through auxiliary variables. I don't know if this is reasonable. 
If I identify a cell, and I think I can do a small loop for all the Gauss points in the cell, is there any way to get an intermediate variable value on the Gauss points?

**Reply by xiaojbing** [reply]
When I tried unstructured grids earlier, by outputting element numbers on the screen, I found that they were not sequential, or in order. Can I have all the element numbers(As far as I know, _current_elem->id() corresponds to the number of the previous cell, which is just a number)?

**Reply by GiudGiud** [reply]
The _qp values from the current cell are all available in elemental loops

_current_elem is the id of the current cell. 

You can loop over elements. Making a loop over the ids is not the way to go

**Comment by hugary1995** [comment]
Take a look at how we compute the element-averaged deformation gradient here: https://github.com/idaholab/moose/blob/1cd91e61ff5b6c9d14ef83a94b6ae512e8a3f546/modules/tensor_mechanics/src/materials/lagrangian/ComputeLagrangianStrainBase.C#L187-L188

The utility that accepts a lambda or a functor in StabilizationUtils should be helpful.

**Comment by friedmud** [comment]
As @hugary1995 shows, it is possible to do this within a `Material`.  For Kernels and other objects... I would recommend using the Auxiliary system.  An Auxkernel that just does `return _coupled_value[_qp]` that fills a CONSTANT MONOMIAL field will compute the element average of the coupled field.  You could then couple that into whatever you want.
